File: src/main/java/com/superpixel/advokit/mapper/JsonContentMapperBuilder.java
package com.superpixel.advokit.mapper;

import java.util.Map;

import com.superpixel.advokit.json.lift.JValueMapper;

public class JsonContentMapperBuilder {

	private Map<String, String> pathMapping;
	private Map<String, String> inclusions;
	
	/***
	 * Specifies the transformation that the json content should go through.
	 * Map keys should be the 'to' path and map values should be the 'from' path.
	 * 
	 * Path format follows javascript object notation. More details can be found in the JPath class file.
	 * 
	 * @param pathMapping
	 * @return
	 */
	public JsonContentMapperBuilder withPathMapping(Map<String, String> pathMapping) {
		this.pathMapping = pathMapping;
		return this;
	}
	
	/***
	 * Specifies inclusions for the content.
	 * Map keys should be inclusion ids and values should be JSON.
	 * 
	 * @param inclusions
	 * @return
	 */
	public JsonContentMapperBuilder withInclusions(Map<String, String> inclusions) {
		this.inclusions = inclusions;
		return this;
	}
	
	public <T> JsonContentMapper<T> build(Class<T> targetClass) {
		JValueMapper.apply(targetClass, JPathPair.fromMap(pathMapping), inclusions)
	}
}

File: src/main/scala/com/superpixel/advokit/mapper/JsonContentMapperBuilder.scala
package com.superpixel.advokit.mapper

import collection.JavaConverters._
import com.superpixel.advokit.json.pathing.JPathPair
import com.superpixel.advokit.json.lift.JValueMapper

class JsonContentMapperBuilder {

  var inclusions: Option[collection.mutable.Map[String, String]] = None
  var pathMappings: Option[collection.mutable.Map[String, String]] = None
  
  def withPathMappings(mapping: java.util.Map[java.lang.String, java.lang.String]): JsonContentMapperBuilder = {
    this.pathMappings = Some(mapping.asScala)
    this
  }
  
  def withInclusions(inclusions: java.util.Map[java.lang.String, java.lang.String]): JsonContentMapperBuilder = {
    this.inclusions = Some(inclusions.asScala)
    this
  }
  
  def build[T](targetClass: Class[T]): JsonContentMapper[T] = {
    
    implicit val m: Manifest[T] = Manifest.classType(targetClass)
    
    val imInclusions: collection.immutable.Map[String, String] = this.inclusions match {
      case None => collection.immutable.Map()
      case Some(inc) => inc.toMap
    }
    val imMapping: collection.immutable.Set[JPathPair] = this.pathMappings match {
      case None => collection.immutable.Set()
      case Some(pM) => {
        val entrySet: collection.immutable.Set[(String, String)] = pM.toSet
        entrySet.map { case (to:String , from:String) => JPathPair.fromStrings(to, from) }
      }
    }
    JValueMapper.forTargetClass(targetClass, imMapping, imInclusions)
  }
}